2019-03-Autumn/1_linked_lists/README.md

```diff
@@ -70,21 +108,124 @@ TODO :bug:
 
 TODO :bug:
 
-### 2. PROBLEM 2 TODO :bug:
-
-Source: TODO :bug:
-
-#### Naive/Simple Solution
-
-TODO :bug:
-
-#### Optimal Solution
-
-TODO :bug:
-
-#### Testing The Solutions OR Driver For Solution
-
-TODO :bug:
+### 2. Loop Detection
+
+Source: CTCI 2.8
+
+#### Solution for Detect Loop
+
+By using the *slow pointer and fast pointer* (or *tortoise and hare*) approach,
+we can guarantee that in the case which there is a loop, the 2 pointers will 
+meet at some node in the loop.
+
+```c++
+struct Node {
+    int data;
+    Node* next;
+};
+
+bool detectLoop(Node* head) {
+    Node* fast = head;
+    Node* slow = head;
+    while (fast != nullptr && fast->next != nullptr) {
+        fast = fast->next->next;
+        slow = slow->next;
+        if (fast == slow) {
+            return true;
+        }
+    }
+    // fast hit nullptr, no loop
+    return false;
+}
+```
+
+#### Solution for Start of Loop (Intro)
+
+In the previous solution, we were able to detect the existence of a loop. Now, 
+to find where the loop starts, one technique is to count the number of nodes in 
+the loop. Since we know if we have a loop, the `n`th node in the loop will be 
+the same as the (`n` + `sizeOfLoop`)th node. Having that information, we can 
```

Comment:
> Since we know if we have a loop, the `n`th node in the loop will be the same as the (`n` + `sizeOfLoop`)th node.

I find this wording to be a little confusing because at first glance it is unclear whether it is `n` from `head` or `n` from the start of the list's cycle. Try ``Since we know we have a loop, we know that for any node `n` in the list where `n` is also in the loop, `n` will be the same node as the (`n` + `sizeOfLoop`)th node.`` or something along those lines. This isn't perfect either, but be careful using `n` where it isn't already clearly defined; specifying whether it is the length of the dangling front section of the list, some arbitrary point in the list, or the size of the loop will help clarify what you mean.